src/feature/hs/hs_service.c

```diff
@@ -3042,13 +3042,85 @@ service_desc_hsdirs_changed(const hs_service_t *service,
   return should_reupload;
 }
 
+/** These are all the reasons why a descriptor upload can't occur. We use
+ * those to log the reason properly with the right rate limiting and for the
+ * right descriptor. */
+typedef enum {
+  LOG_DESC_UPLOAD_REASON_MISSING_IPS = 0,
+  LOG_DESC_UPLOAD_REASON_IP_NOT_ESTABLISHED = 1,
+  LOG_DESC_UPLOAD_REASON_NOT_TIME = 2,
+  LOG_DESC_UPLOAD_REASON_NO_LIVE_CONSENSUS = 3,
+  LOG_DESC_UPLOAD_REASON_NO_DIRINFO = 4,
+
+  /* Always +1 of the last value. This is used to allocate the static array of
+   * all rate limiting objects. */
+  LOG_DESC_UPLOAD_REASON_NUM_ = 5,
```

Comment:
I don't like how this "size of enum" object is also part of the enum. It's not a reason itself, so I don't think it should be in the reason enum for semantic reasons. Perhaps it can be defined separately as a #define using `sizeof(log_desc_upload_reason_t)`?

Also I'm not sure I like the trailing `_`. What does it signify?

Perhaps instead of `NUM` we could use `MAX`? I think that's thje unwritten convention in the rest of the code. I think that would be better since there are not actually 5 reasons; they are 4.

Reply:
Renamed this and set it as a `#define` instead.